### PR TITLE
[SystemZ][z/OS] Add goffdumper/llvm-readobj tools

### PR DESCRIPTION
**Pull request by ysyeda**
This PR adds the GOFFDumper. It contains a test which uses llvm-readobj to check the symbols of the provided GOFF object file.

### CODE REVIEW

File: llvm/tools/llvm-readobj/GOFFDumper.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+using namespace llvm::object;
+
+namespace {
+
+class GOFFDumper : public ObjDumper {
+public:
+  GOFFDumper(const GOFFObjectFile *Obj, ScopedPrinter &Writer)
+      : ObjDumper(Writer, Obj->getFileName()), Obj(Obj) {}
+
+  void printFileHeaders() override {}
```
Comment (jh7370) [inline llvm/tools/llvm-readobj/GOFFDumper.cpp:16]: I don't know the GOFF format at all, but will it make sense to support this (and `printUnwindInfo` and `printStackMap`) in the future? If so, I would recommend adding a warning to these methods saying that they are unimplemented.

File: llvm/tools/llvm-readobj/GOFFDumper.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+};
+
+} // End anonymous namespace
+
+namespace llvm {
+std::unique_ptr<ObjDumper> createGOFFDumper(const object::GOFFObjectFile &Obj,
+                                            ScopedPrinter &Writer) {
+  return std::make_unique<GOFFDumper>(&Obj, Writer);
+}
+
```
Comment (jh7370) [inline llvm/tools/llvm-readobj/GOFFDumper.cpp:39]: Nit: be consistent with your start/end of your namespacing with regards to blank lines.

File: llvm/tools/llvm-readobj/GOFFDumper.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+private:
+  void printSymbol(const SymbolRef &Sym);
+
+  const GOFFObjectFile *Obj;
+};
+
+} // End anonymous namespace
+
+namespace llvm {
+std::unique_ptr<ObjDumper> createGOFFDumper(const object::GOFFObjectFile &Obj,
```
Comment (jh7370) [inline llvm/tools/llvm-readobj/GOFFDumper.cpp:35]: This should use namespace qualifiers, per the coding standard.

File: llvm/tools/llvm-readobj/GOFFDumper.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+void GOFFDumper::printSymbols(bool) {
+  ListScope Group(W, "Symbols");
+
+  for (const SymbolRef &Symbol : Obj->symbols())
+    printSymbol(Symbol);
+}
+
+void GOFFDumper::printDynamicSymbols() { ListScope Group(W, "DynamicSymbols"); }
+
+void GOFFDumper::printRelocations() {}
```
Comment (jh7370) [inline llvm/tools/llvm-readobj/GOFFDumper.cpp:94]: Not sure why you've provided these stub definitions out of line unlike the ones I commented on above?

**Review by jh7370 — COMMENTED** [review]
You should really try to avoid canned object files in the repository, because thwey are opaque and it's not clear what in them us important for testing them. They also don't play well with source control (you can't diff a binary file). A better approach would be to generate the object at test time. From most preferable to least, you could use a) yaml2obj, b) assembly and llvm-mc, or c) IR and llc. I expect yaml2obj doesn't support GOFF yet, but adding that support will be highly worthwhile if you plan on adding readobj and other tool support, because it allows you fine-grained control of your test inputs, allowing you to more easily test edge cases (especially for error handling).

You should also break this PR into multiple. The first would do nothing except add the bare framework for GOFF files, and the rest should add one feature (symbol dumping, section dumping etc) at a time.
